### PR TITLE
Features cumsum & cumprod

### PR DESCRIPTION
**Pull request by mtar**
## Description

Include a summary of the change/s.
Please also include relevant motivation and context. List any dependencies that are required for this change.

implementation of cumsum and cumprod functions. numpy's axis=None is not supported.

Issue/s resolved: #243 

## Changes proposed:
- add cumsum
- add cumprod
- cumproduct as alias to cumprod

## Type of change

Remove irrelevant options:
- New feature (non-breaking change which adds functionality)
- Documentation update

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
yes / no

### CODE REVIEW

File: heat/core/arithmetics.py [hunk 1]
```diff
@@ -197,6 +203,66 @@ def bitwise_xor(t1, t2):
...
+        cprod *= recv
+
+    return factories.array(cprod, dtype=a.dtype, is_split=a.split, device=a.device)
+
+
+# Alias support
+cumproduct = cumprod
+
+
+def cumsum(a, axis):
```
Comment (mtar) [inline heat/core/arithmetics.py:238]: Docstring & additional parameters

File: heat/core/tests/test_arithmetics.py [hunk 2]
```diff
@@ -161,6 +161,74 @@ def test_bitwise_xor(self):
         with self.assertRaises(TypeError):
             ht.bitwise_xor(self.an_int_scalar, self.a_scalar)
 
+    def test_cumprod(self):
```
Comment (mtar) [inline heat/core/tests/test_arithmetics.py:164]: docstring and additional arguments

File: heat/core/arithmetics.py [hunk 3]
```diff
@@ -197,6 +203,66 @@ def bitwise_xor(t1, t2):
     return operations.__binary_op(torch.Tensor.__xor__, t1, t2)
 
 
+def cumprod(a, axis):
+    if axis is None:
```
Comment (coquelin77) [inline heat/core/arithmetics.py:207]: docs

File: heat/core/tests/test_arithmetics.py (hunk 2 above)
Comment (coquelin77) [reply heat/core/tests/test_arithmetics.py:164]: not needed here

File: heat/core/tests/test_arithmetics.py [hunk 4]
```diff
@@ -161,6 +161,74 @@ def test_bitwise_xor(self):
...
+
+        a = ht.full((4, 2), 2, dtype=ht.int64, split=0, device=ht_device)
+        result = ht.array(
+            [[2, 2], [4, 4], [8, 8], [16, 16]], dtype=ht.int64, split=0, device=ht_device
+        )
+
+        cprod = ht.cumprod(a, 0)
+        self.assertTrue(ht.equal(cprod, result))
+
+        a = ht.full((2, 4, 2), 2, dtype=ht.float32, split=1, device=ht_device)
```
Comment (coquelin77) [inline heat/core/tests/test_arithmetics.py:184]: please test up to 3 dims

File: heat/core/arithmetics.py [hunk 5]
```diff
@@ -197,6 +203,66 @@ def bitwise_xor(t1, t2):
...
+        if cprod.numel() > 0:
+            indices = -1
+            Ni, Nk = cprod.shape[:axis], cprod.shape[axis + 1 :]
+            for ii in np.ndindex(Ni):
+                for kk in np.ndindex(Nk):
+                    send[ii + np.s_[...,] + kk] = cprod[ii + np.s_[:,] + kk][indices]
+
+        recv = torch.ones(cprod.shape[:axis] + cprod.shape[axis + 1 :], dtype=send.dtype)
+
+        a.comm.Exscan(send, recv, MPI.PROD)
```
Comment (coquelin77) [inline heat/core/arithmetics.py:226]: @krajsek does this work with the upcoming AD stuff?

File: heat/core/tests/test_arithmetics.py (hunk 2 above)
Comment (mtar) [reply heat/core/tests/test_arithmetics.py:164]: Wrong line

File: heat/core/arithmetics.py (hunk 3 above)
Comment (mtar) [reply heat/core/arithmetics.py:207]: this is the line I meant and additional arguments

File: heat/core/arithmetics.py [hunk 6]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+    Returns
+    -------
+    cumprod : DNDarray
+        A new array holding the result is returned unless `out` is
+        specified, in which case a reference to out is returned.
+    """
+
+    if axis is None:
+        raise NotImplementedError("axis = None is not supported")
+    if not isinstance(axis, int):
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:236]: This should be done with stide_tricks.sanitize_axis right?

File: heat/core/arithmetics.py [hunk 7]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+        Alternative output array in which to place the result. It must
+        have the same shape and buffer length as the expected output
+        but the type of the resulting values will be cast if necessary.
+
+    Returns
+    -------
+    cumprod : DNDarray
+        A new array holding the result is returned unless `out` is
+        specified, in which case a reference to out is returned.
+    """
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:232]: Add Raises section for exceptions that are thrown from within this function

File: heat/core/arithmetics.py [hunk 8]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+        raise NotImplementedError("axis = None is not supported")
+    if not isinstance(axis, int):
+        raise RuntimeError("axis parameter got an invalid argument {}".format(axis))
+
+    if dtype is not None:
+        dtype = types.canonical_heat_type(dtype)
+
+    if out is not None:
+        if out.shape != a.shape:
+            raise RuntimeError(
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:244]: Should be ValueError

File: heat/core/arithmetics.py (hunk 5 above)
Comment (Markus-Goetz) [reply heat/core/arithmetics.py:276]: Not yet.

File: heat/core/arithmetics.py [hunk 9]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+    Returns
+    -------
+    cumsum_along_axis : DNDarray.
+        A new array holding the result is returned unless `out` is
+        specified, in which case a reference to `out` is returned. The
+        result has the same size as `a`, and the same shape as `a` if
+        `axis` is not None or `a` is a 1-d array.
+
+    """
+    if axis is None:
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:324]: sanitize_axis

File: heat/core/arithmetics.py [hunk 10]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+        raise NotImplementedError("axis = None is not supported")
+    if not isinstance(axis, int):
+        raise RuntimeError("axis parameter got an invalid argument {}".format(axis))
+
+    if dtype is not None:
+        dtype = types.canonical_heat_type(dtype)
+
+    if out is not None:
+        if out.shape != a.shape:
+            raise RuntimeError(
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:334]: ValueError

File: heat/core/arithmetics.py [hunk 11]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+        precision less than that of the default platform integer.  In
+        that case, the default platform integer is used.
+    out : DNDarray, optional
+        Alternative output array in which to place the result. It must
+        have the same shape and buffer length as the expected output
+        but the type will be cast if necessary. See `doc.ufuncs`
+        (Section "Output arguments") for more details.
+
+    Returns
+    -------
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:316]: Documentation

File: heat/core/arithmetics.py [hunk 12]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+        if csum.numel() > 0:
+            indices = -1
+            Ni, Nk = csum.shape[:axis], csum.shape[axis + 1 :]
+            for ii in np.ndindex(Ni):
+                for kk in np.ndindex(Nk):
+                    send[ii + np.s_[...,] + kk] = csum[ii + np.s_[:,] + kk][indices]
+
+        recv = torch.zeros(csum.shape[:axis] + csum.shape[axis + 1 :], dtype=send.dtype)
+
+        a.comm.Exscan(send, recv, MPI.SUM)
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:366]: More or less identical code as above. Could you please introduce a __cum_op() call analogous to __reduce_op() in operations.py and make use of it?

File: heat/core/arithmetics.py [hunk 13]
```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
...
+        dtype=None if dtype is None else dtype.torch_type(),
+    )
+
+    if a.split is not None and axis == a.split:
+        send = torch.zeros(csum.shape[:axis] + csum.shape[axis + 1 :], dtype=csum.dtype)
+
+        if csum.numel() > 0:
+            indices = -1
+            Ni, Nk = csum.shape[:axis], csum.shape[axis + 1 :]
+            for ii in np.ndindex(Ni):
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py:360]: Can we somehow manage to do this without the Python loops via views? Performance on them should be slow.

Comment (mtar) [reply heat/core/arithmetics.py]: I replaced it with torch.index_select()

File: heat/core/arithmetics.py (hunk 12 above)
Comment (mtar) [reply heat/core/arithmetics.py]: function added

File: heat/core/arithmetics.py (hunk 7 above)
Comment (mtar) [reply heat/core/arithmetics.py]: Shall the raises section only be in  __cum_op() or also here?

File: heat/core/arithmetics.py (hunk 6 above)
Comment (mtar) [reply heat/core/arithmetics.py]: Done
